pkgs/os-specific/linux/zfs/default.nix

```diff
@@ -68,7 +68,7 @@ let
           # And if it's enabled by default, only change that if we explicitly disable python to remove python from the closure
           nfs-utils.override (old: { enablePython = old.enablePython or true && enablePython; })
         }/bin/exportfs"
-        substituteInPlace ./lib/libshare/smb.h        --replace "/usr/bin/net"            "${samba}/bin/net"
+        substituteInPlace ./lib/libshare/smb.h        --replace "/usr/bin/net"            "/run/current-system/sw/bin/net"
```

Comment:
This is not what I meant. I meant to say to make it optional, not impure.

Reply:
But this variant works and it is not necessary to recompile. If enable the samba service, the `net` command is installed in the `/run/current-system/sw/bin` folder.

Reply:
Using `${samba}/bin/net` is also a bad idea because there is `services.samba.package` which is not necessarily the same as `samba`. To prevent diverging version the easiest way is to rely on `/run/current-system/sw/bin/net`.

Reply:
but using anything from ``/run/current-system/sw/`` is impure and causing problems when glibc differs while updating and is already causing enough problems with graphics drivers. Do we have another option at hand?

Reply:
I am not sure either where that is defined. Somewhere the zfs package is drawned into and there we need to override it.

Reply:
You can't easily override the kernel module because the kernel is set (or at least can be set) by the user via `boot.kernelPackages`.

Reply:
I'm for the current variant :)

Reply:
I would prefer `/usr/bin/env net` but that's unfortunately not possible because arguments to `execv` cannot have spaces.

Reply:
If complete this variant? I don't know how to read variable.
```diff
diff --git a/nixos/modules/services/network-filesystems/samba.nix b/nixos/modules/services/network-filesystems/samba.nix
index 992f948e8cd..873793b2074 100644
--- a/nixos/modules/services/network-filesystems/samba.nix
+++ b/nixos/modules/services/network-filesystems/samba.nix
@@ -244,6 +244,7 @@ in

         security.pam.services.samba = {};
         environment.systemPackages = [ cfg.package ];
+        environment.variables = { LOCATE_SAMBA_NET = "${cfg.package}/bin/net"; };

         networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ 139 445 ];
         networking.firewall.allowedUDPPorts = mkIf cfg.openFirewall [ 137 138 ];
diff --git a/pkgs/os-specific/linux/zfs/default.nix b/pkgs/os-specific/linux/zfs/default.nix
index 5a5fd2d19a8..9025ba942ca 100644
--- a/pkgs/os-specific/linux/zfs/default.nix
+++ b/pkgs/os-specific/linux/zfs/default.nix
@@ -68,7 +68,7 @@ let
           # And if it's enabled by default, only change that if we explicitly disable python to remove python from the closure
           nfs-utils.override (old: { enablePython = old.enablePython or true && enablePython; })
         }/bin/exportfs"
-        substituteInPlace ./lib/libshare/smb.h        --replace "/usr/bin/net"            "${samba}/bin/net"
+        substituteInPlace ./lib/libshare/smb.h        --replace '"/usr/bin/net"'          'getenv("LOCATE_SAMBA_NET")'
         substituteInPlace ./config/user-systemd.m4    --replace "/usr/lib/modules-load.d" "$out/etc/modules-load.d"
         substituteInPlace ./config/zfs-build.m4       --replace "\$sysconfdir/init.d"     "$out/etc/init.d" \
                                                       --replace "/etc/default"            "$out/etc/default"

```

Reply:
And if `getenv` returns `NULL` everything will blow up. Please don't do this, this is by far the worst of all options.

Reply:
with `getenv("LOCATE_SAMBA_NET") NULL` doesn't work either - error build.

Reply:
I mean that you need
```c
const char *samba_net = getenv("LOCATE_SAMBA_NET");
if (samba_net == NULL) {
    samba_net = "/run/current-system/sw/bin/net"; // or some other sensible default
}
```
but you can't actually do this because what is being patched here is the result of a macro expansion and in the context where the macro is used it can only expand to a single statement. Better leave it at `"/run/current-system/sw/bin/net"` for now and raise an issue upstream, asking to make the Samba executable configurable.

Reply:
@hmenke created issue.

Reply:
If you're going to have it be impure (which I don't think you should), why not have it just run it from PATH?

Reply:
Replace "/usr/bin/net" to "net"?
I'll try to check this variant.

Reply:
@alyssais `/usr/bin/net` is passed to `execv`, so sadly only absolute file paths (or patch ZFS further to use `execvp` or so).

Reply:
Not workin this variant.

Reply:
since `execv` needs an absolute path could we just make a wrapper script and absolutely point to that?